### PR TITLE
getJsonObject: improve performance for new version of getJsonObject

### LINKED ISSUE #1907: [FEA] [follow-up] getJsonObject: improve performance for new version of getJsonObject
This is from: https://github.com/NVIDIA/spark-rapids-jni/pull/1893

Due to limited time, #1893 only focus on the correctness of new version of getJsonObject,
use this follow-up issue to do perf improvement.

maybe:
- optimize the stack memory.
- optimize push/pop stask, use peak may be more efficient.

### PR DESCRIPTION
**Pull request by res-life**
closes #1907

Improve performance for new version of getJsonObject.
Fix a redefined variable

### perf test
Tested ByteDance data.
The first line is cuDF impl.

|After commit|time(1st,2nd,3rd)| time / original cuDF impl (less is better)|
|:-|:-|:-|
|original cuDF impl, branch 24.02| 2183 ms, 778 ms, 756 ms| base 1.0|
|new version of get-json-obj in branch 24.04| 21230 ms, 19887 ms, 20150 ms|16.482916330374|
|Refactor: Move device code in get_json_object to cu or cuh files|21278 ms, 19917 ms, 20366 ms|16.5620123755717|
|Refactor/simplify json generator|6113 ms, 4734 ms, 4690 ms|4.17998385794996|
|Remove purge non-empty nulls step|6064 ms, 4747 ms, 4753 ms|4.18724778046812|
|Refactor josn parser: remove useless variables|6013 ms, 4687 ms, 4752 ms|4.15711595372612|
|Use 64 bits to store JSON nest depth context to save memory|6086 ms, 4695 ms, 4709 ms|4.16733925208502|
|Refine push/pop/peek logic|2850 ms, 1570 ms, 1595 ms|1.6182405165456|
|Refactor: use less functions; change max path length from 32 to 8|2716, 1584, 1515|1.56443368307775|

Signed-off-by: Chong Gao <res_life@163.com>

### CODE REVIEW

**Comment by res-life** [comment]
Original cuDF uses time about 700ms when testing ByteDance data.

|commit|time|
|:-:|:-:|
|init new version|15s|
|json_nest_depth = 8; max_path=4|5s|
|use bits in int64 to save json nest depth; json_nest_depth = 64; max_path=8|5s|
|Refactor: reduce push/pop; json_nest_depth = 64; max_path=8 |1.7s|
|Refactor: reduce construct context; json_nest_depth = 64; max_path=8 |1.6s|
|Refactor generator |1.6s|
|allocate uvector<context> |avg 700ms|

After some commits, we can get 700ms as original cuDF impl.

Note: currently meets a error in QA's env:
```
ai.rapids.cudf.CudaFatalException: parallel_for: failed to synchronize: cudaErrorIllegalAddress: an illegal memory access was encountered
```

**Comment by res-life** [comment]
Please refer to the last commit, it fix nvbug: Fix nvbug: get-json-obj get incorrect result on some GPUs(H100, V100).
Seems it's related to `thrust::tie`, on some GPU(H100 V100) CUDA(12.x), tie generate whong result(empty output). And on T4, does not occur. It's weird issue.
After changed to tie a structure instead of `cudf::size_type`, then generate correct answer.
Currently do not know why this fix works.

**Comment by ttnghia** [comment]
Please try reverting `get_json_object.cu` in the latest commit and apply this:
```diff
@@ -1238,21 +1240,20 @@ __launch_bounds__(block_size) CUDF_KERNEL
   while (tid < col.size()) {
     bool is_valid               = false;
     cudf::string_view const str = col.element<cudf::string_view>(tid);
-    cudf::size_type output_size = 0;
     if (str.size_bytes() > 0) {
       char* dst = out_buf != nullptr ? out_buf + output_offsets[tid] : nullptr;
       size_t const dst_size =
         out_buf != nullptr ? output_offsets[tid + 1] - output_offsets[tid] : 0;
 
       // process one single row
-      auto [result, output_size] = get_json_object_single(
+      auto const [result, output_size] = get_json_object_single(
         str.data(), str.size_bytes(), path_commands_ptr, path_commands_size, dst, dst_size);
       if (result) { is_valid = true; }
-    }
 
-    // filled in only during the precompute step. during the compute step, the
-    // offsets are fed back in so we do -not- want to write them out
-    if (out_buf == nullptr) { d_sizes[tid] = output_size; }
+      // filled in only during the precompute step. during the compute step, the
+      // offsets are fed back in so we do -not- want to write them out
+      if (out_buf == nullptr) { d_sizes[tid] = static_cast<cudf::size_t>(output_size); }
+    }
```

File: src/main/cpp/src/get_json_object.hpp [hunk 1]
```diff
@@ -13,7 +13,6 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-#pragma once
```
Comment (ttnghia) [inline src/main/cpp/src/get_json_object.hpp]: This should not be reverted.

**Comment by res-life** [comment]
>Please try reverting get_json_object.cu in the latest commit and apply this:

Done, with a little change:

```
    } else {
      // valid JSON length is always greater than 0
      // if `str` size len is zero, output len is 0 and `is_valid` is false
      if (out_buf == nullptr) { d_sizes[tid] = 0; }
    }
```

**Comment by res-life** [comment]
@ttnghia  thanks for your help, please help review again.

**Review by ttnghia — APPROVED** [review]

**Comment by ttnghia** [comment]
Performance change after this PR, compared using https://github.com/NVIDIA/spark-rapids-jni/pull/1952:
```
|  size_bytes  |  max_depth  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |           Diff |   %Diff |  Status  |
|--------------|-------------|------------|-------------|------------|-------------|----------------|---------|----------|
|   1000000    |      2      |   6.654 ms |      10.40% |   5.576 ms |      11.35% |   -1078.003 us | -16.20% |   FAIL   |
|   10000000   |      2      |   7.373 ms |       8.92% |   6.230 ms |      11.22% |   -1143.858 us | -15.51% |   FAIL   |
|  100000000   |      2      | 145.607 ms |       1.81% | 131.515 ms |       1.96% |  -14092.123 us |  -9.68% |   FAIL   |
|  1000000000  |      2      |    1.745 s |       0.67% |    1.558 s |       1.39% | -186955.541 us | -10.72% |   FAIL   |
|   1000000    |      4      |   7.300 ms |       8.87% |   5.836 ms |      12.70% |   -1463.903 us | -20.05% |   FAIL   |
|   10000000   |      4      |   8.247 ms |       7.03% |   6.414 ms |       8.63% |   -1833.158 us | -22.23% |   FAIL   |
|  100000000   |      4      | 159.210 ms |       0.98% | 133.165 ms |       1.58% |  -26044.341 us | -16.36% |   FAIL   |
|  1000000000  |      4      |    1.872 s |       0.92% |    1.518 s |       1.00% | -354327.340 us | -18.93% |   FAIL   |
|   1000000    |      6      |   7.639 ms |       9.03% |   5.905 ms |      10.32% |   -1734.084 us | -22.70% |   FAIL   |
|   10000000   |      6      |   8.646 ms |       8.02% |   6.346 ms |       7.87% |   -2300.449 us | -26.61% |   FAIL   |
|  100000000   |      6      | 165.838 ms |       1.92% | 131.152 ms |       1.16% |  -34686.127 us | -20.92% |   FAIL   |
|  1000000000  |      6      |    1.925 s |       1.07% |    1.581 s |       0.68% | -344136.035 us | -17.87% |   FAIL   |
|   1000000    |      8      |   8.109 ms |       6.55% |   6.090 ms |      10.00% |   -2019.178 us | -24.90% |   FAIL   |
|   10000000   |      8      |   9.414 ms |       5.90% |   6.598 ms |       8.01% |   -2815.621 us | -29.91% |   FAIL   |
|  100000000   |      8      | 177.391 ms |       1.20% | 134.754 ms |       1.20% |  -42636.822 us | -24.04% |   FAIL   |
|  1000000000  |      8      |    2.034 s |       0.54% |    1.576 s |       0.71% | -457780.972 us | -22.51% |   FAIL   |
```

**Pull request merged**